lib/propolis/src/accessors.rs

```diff
@@ -12,54 +12,84 @@
 //! contained emulation should fail any DMA accesses.
 
 use std::collections::{BTreeMap, BTreeSet, VecDeque};
+use std::ffi::c_void;
 use std::marker::PhantomData;
 use std::sync::{Arc, Mutex, MutexGuard, Weak};
 
 use crate::vmm::{MemCtx, VmmHdl};
 
-type AccId = usize;
-
-const ID_NULL: AccId = 0;
-const ID_ROOT: AccId = 1;
+/// Key type for identifying nodes referenced by `Tree`.
+#[derive(Ord, PartialOrd, Eq, PartialEq, Debug, Copy, Clone)]
+#[repr(transparent)]
+pub struct NodeKey(*const Node<c_void>);
+impl NodeKey {
+    const NULL: Self = Self(std::ptr::null());
+}
```

Comment:
IMO, it might be nicer to represent this as `NonNull<Node<c_void>>`, allowing us to use `Option<NodeKey>` rather than treating `NodeKey` as nullable (and using `NonNull` means that `Option<NodeKey>` should be represented the same as `NodeKey(ptr::null())` due to niche optimization). 

This way, code that touches a nullable `NodeKey` (the parent ID) is forced to handle the null case, while the compiler ensures that other `NodeKey`s that are not the parent key are never null.

Reply:
I waffled a bit on this while writing it.  It makes some accesses a bit more annoying, and further suggests to one examining `NodeKey` that the pointer it holds might actually be useful for something (beyond a unique ID).  The help from the compiler is probably worth it, though.